File: DoubtClass-7(MedianOfRowWiseSortedMatrix-NonRepeatingElement-SearchRotatedSortedArray-PascalsTriangle)/SearchRotatedSortedArray.cpp
int calculateMedianOfMatrix(vector<vector<int> > &matrix) {
    // add your logic here
	int n = matrix.size();
	int m = matrix[0].size();
	int mine = INT_MAX;
	int maxe = INT_MIN;
	for(int i=0;i<n;i++) {
		mine = min(mine,matrix[i][0]);
		maxe = max(maxe,matrix[i][m-1]);
	}
	int target = (n*m + 1)/2;
		int ans = mine;
		while(mine <= maxe) {
			int mid = (mine + maxe)/2;
			int count = 0;
			for(int i=0;i<n;i++) {
				int index = upper_bound(matrix[i].begin(),matrix[i].end(),mid) - matrix[i].begin();
				count += index;
			}
			if(count < target) {
				mine = mid+1;
			}
			else {
				ans = mid;
				maxe = mid-1;
			}
		}
		return ans;
}

File: DoubtClass-7(MedianOfRowWiseSortedMatrix-NonRepeatingElement-SearchRotatedSortedArray-PascalsTriangle)/SearchRotatedSortedArray.java
class Solution {
// 	private int search(int arr[], int l, int h, int key) {
//         if (l > h)
//             return -1;
 
//         int mid = (l + h) / 2;
//         if (arr[mid] == key)
//             return mid;
//         if (arr[l] <= arr[mid]) {
//             if (key >= arr[l] && key <= arr[mid])
//                 return search(arr, l, mid - 1, key);
//             return search(arr, mid + 1, h, key);
//         }
//         if (key >= arr[mid] && key <= arr[h])
//             return search(arr, mid + 1, h, key);
//         return search(arr, l, mid - 1, key);
//     }
	int search(int arr[],int low,int high,int key) {
		if(low > high) {
			return -1;
		}
		int mid = (low + high)/2;
		if(arr[mid] == key) {
			return mid;
		}
		if(arr[low] <= arr[mid]) {
			if(key >= arr[low] && key <= arr[mid]) {
				return search(arr,low,mid-1,key);
			}
			else {
				return search(arr,mid+1,high,key);
			}
		}
		else {
			if(key >= arr[mid] && key <= arr[high]) {
				return search(arr,mid+1,high,key);
			}
			else {
				return search(arr,low,mid-1,key);
			}
		}
	}
	int getElementIndex(int[] array, int target) {
	    // add your logic here
		return search(array,0,array.length-1,target);
	}
}